diff --git a/Eigen/Core b/Eigen/Core
index b923b8c..874be88 100644
--- a/Eigen/Core
+++ b/Eigen/Core
@@ -305,6 +305,8 @@
 /** \brief Namespace containing all symbols from the %Eigen library. */
 namespace Eigen {
 
+using std::real;
+
 inline static const char *SimdInstructionSetsInUse(void) {
 #if defined(EIGEN_VECTORIZE_AVX512)
   return "AVX512, FMA, AVX2, AVX, SSE, SSE2, SSE3, SSSE3, SSE4.1, SSE4.2";
